ContactMangerUI [Step1, 2] june & Joo

ohdair opened the pull request:
안녕하세요 소대(@zdodev) 😎
이번에 연락처 관리 프로그램(UI Ver) 프로젝트를 진행하게 된 Joo, june입니다!
STEP 1, 2 규모가 작아서 묶어서 PR 보내도 되겠다고 생각했는데 흔쾌히 허락해주셔서 감사합니다 🙂
이번 프로젝트에서는 UITableViewCell을 xib로 구현해봤어요! 관련된 고민과 질문사항 아래에 적어놓았는데 소대의 의견을 듣고 싶습니다! 이번 프로젝트도 잘 부탁드릴게요 🙏

## 🚀 STEP 1 - iOS App Target 추가

### ✔️ 구현사항
- [x] 첫 주차의 연락처 관리 프로그램의 Xcode 프로젝트를 기반으로 시작합니다.
- [x] 기존의 연락처 관리 프로그램 프로젝트에 UIKit App Target을 추가합니다
    - Target 이름 : ContactManagerUI
    - Storyboard를 사용하며, Swift 언어를 사용합니다
- [x] ContactManager 타깃의 주요 파일을 ContactManagerUI의 타깃 멤버십 파일로 추가합니다
    - main.swift 제외
    - 즉, 프로그램 구동의 주요 타입과 기능은 main.swift에 존재하면 안된다는 뜻입니다

### 🔑 핵심경험
- Xcode의 Target 개념 이해
- Xcode의 새로운 Target 생성
- Target Membership 개념 이해

---

## 🚀 STEP 2 - 연락처 목록 구현
<img width="50%" src="https://user-images.githubusercontent.com/22979718/216210634-edd03898-f5d0-4afe-9dcf-dfa702e04b71.png">

### ✔️ 구현사항
- [x] TableView를 사용하여 연락처 목록 화면에 표시
- [x] 각 행의 cell은 subtitle style을 사용
- [x] 각 행(row)에 항목을 표시
    - 이름
    - 나이
    - 연락처

### 🔑 핵심경험
- Table View의 기초구현

### 🤔 고민 및 질문
- UITableViewCell의 구성하는 방법으로는 storyboard, programmatically, xib가 있는 것을 확인하였고 무엇을 사용할까 고민했습니다. 우선은 사용에 대한 초점보다는 공부에 목적을 두었기에 xib를 활용하여 cell을 만들었습니다. 소대라면 어떠한 방식으로 구현하셨을지 여쭤봐도 될까요?
- subtitle style의 cell을 사용할 때,title과 subtitle의 변수명을 그대로 사용하였습니다. 이러한 방식이 괜찮은지 궁금합니다.

zhunhe committed:
chore: Create project

zhunhe committed:
chore: Create ContactManagerUI target

ohdair committed:
chore: .gitignore 추가

zhunhe committed:
chore: xcuserdata 관련 파일 삭제

ohdair committed:
feat: UITableView과 UITableVIewCell 구현

- UITableViewCell은 xib을 이용하여 구현
- UITableViewCell에서 subtitle style과 Disclosure Indicator 사용

zhunhe committed:
chore: 더미 데이터를 이용한 연락처 목록 표시

zdodev commented on ContactManager/ContactManager/main.swift:
```diff
@@ -0,0 +1,12 @@
+
+var manager = ContactManager()
+
+manager.runProgram()
```
좋습니다!👍

zdodev commented on ContactManager/ContactManager/Contact.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+struct Contact: Hashable {
+    var name: String
+    var age: UInt
+    var phoneNumber: String
```
프로퍼티를 `var` 로 선언하신 이유가 궁금합니다!

zhunhe replied:
추후에 수정될 수도 있을거라 예상하고 `var`로 구현했었는데 불필요하네요 ! `let` 으로 수정했습니다 !
Joo가 요즘 리팩터링 책을 보고 있는데 YAGNI(You Aren't Going to Need It, 필요한 작업만 해라)에 해당하는 내용인 것 같아요. 🤪

zdodev commented on ContactManager/ContactManager/Error.swift:
```diff
@@ -0,0 +1,43 @@
+
+import Foundation
+
+enum ContactManagerError: Error {
+    case invalidCommand
+    case invalidInput
+    case emptyInput
+    case invalidName
+    case invalidAge
+    case invalidPhoneNumber
+    case existContact
+    case excludeContact(name: String)
+}
+
+extension ContactManagerError: LocalizedError {
```
`ContactManagerError` 타입에 `Error` 프로토콜도 채택하고 `LocalizedError` 프로토콜도 채택한 이유가 궁금합니다!

zhunhe replied:
`LocalizedError`는 `errorDescription` 사용을 위해 채택했고,
`LocalizedError` 내부에서 `Error` 프로토콜을 채택하고 있어 중복이라 판단하여 삭제했습니다 : )

zdodev commented on ContactManager/ContactManagerUI/Controller/ViewController.swift:
```diff
@@ -0,0 +1,61 @@
+final class ViewController: UIViewController {
+
+    @IBOutlet weak var tableView: UITableView!
+    let contactTableViewCell = "ContactTableViewCell"
```
- `internal` 접근 수준으로 구현하신 이유가 있을까요?
- 프로퍼티 이름으로 보아 `String` 타입이 아니라 `UITableViewCell` 과 같은 UI 타입으로 오해할 수 있을 것 같습니다.
상수 문자열을 관리할 수 있는 더 좋은 방법이 있을까요?

zhunhe replied:
접근제어자 설정이 되어있지 않았네요 😅 아래와 같이 수정했습니다 !
```swift
// BEFORE
@IBOutlet weak var tableView: UITableView!

// AFTER
@IBOutlet private weak var tableView: UITableView!
```

zhunhe replied:
`let contactTableViewCell = "ContactTableViewCell"`은 삭제하고 아래와 같이 className으로 구현했습니다 !
이렇게 ContactTableViewCell 내부에서 클래스 이름을 직접 받아오는 방식으로 구현하면 클래스명이 변경되어도 대응가능할거 같아요 😎
```swift
class ContactTableViewCell: UITableViewCell {

    @IBOutlet weak var title: UILabel!
    @IBOutlet weak var subtitle: UILabel!

    static var className: String {
        return String(describing: self)
    }
    ...
```

zdodev commented on ContactManager/ContactManagerUI/Controller/ViewController.swift:
```diff
@@ -0,0 +1,61 @@
+
+extension ViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return contacts.count
```
좋습니다!👍

zdodev commented on ContactManager/ContactManagerUI/Controller/ViewController.swift:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: contactTableViewCell, for: indexPath) as! ContactTableViewCell
```
`dequeueReusableCell` 메서드를 활용해보셨네요!👍
강제 언래핑을 사용한 이유가 궁금합니다!

zhunhe replied:
프로젝트 제약사항에 강제 언래핑을 사용하지 않도록 명시되어있었는데, 사용해버렸네요 ;(
아래와 같이 수정했습니다. ☺️
```swift
// BEFORE
let cell = tableView.dequeueReusableCell(withIdentifier: ContactTableViewCell.className, for: indexPath) as! ContactTableViewCell

// AFTER
guard let cell = tableView.dequeueReusableCell(
    withIdentifier: ContactTableViewCell.className,
    for: indexPath
    ) as? ContactTableViewCell else {
        return UITableViewCell()
    }
```

zdodev commented on ContactManager/ContactManagerUI/Controller/ViewController.swift:
```diff
@@ -0,0 +1,61 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: contactTableViewCell, for: indexPath) as! ContactTableViewCell
+
+        let contact = contacts[indexPath.row]
+        cell.title.text = "\(contact.name)(\(contact.age))"
+        cell.subtitle.text = "\(contact.phoneNumber)"
```
셀의 데이터를 설정해보셨네요!
ViewController 에서 셀의 데이터를 직접 설정하기 보다는 셀에게 맡기는 것은 어떨까요?

zhunhe replied:
말씀해주신대로 셀 클래스 내부에 메서드를 추가하여 셀의 데이터를 설정하도록 만들었습니다 ! ☺️
```swift
extension ContactTableViewCell {
    func configure(contact: Contact) {
        nameAndAge.text = "\(contact.name)(\(contact.age))"
        phoneNumber.text = contact.phoneNumber
    }
}
```

zdodev reviewed (changes requested):
안녕하세요. @ohdair , @zhunhe !

Step 1, 2 구현 잘 진행해주셨습니다!
스텝 요구사항을 잘 지켜주셨네요!👍

질문에 대한 제 의견을 적어보았습니다.

❓UITableViewCell의 구성하는 방법으로는 storyboard, programmatically, xib가 있는 것을 확인하였고 무엇을 사용할까 고민했습니다. 우선은 사용에 대한 초점보다는 공부에 목적을 두었기에 xib를 활용하여 cell을 만들었습니다. 소대라면 어떠한 방식으로 구현하셨을지 여쭤봐도 될까요?

- 개인적으로 programmatically 으로 구현하는 방법을 선호합니다. 하지만 경우에 따라서 다른 방식을 써야할 수 있어서 다양한 방법으로 선언하고 사용하는 방법을 익히는 것이 좋습니다!

❓subtitle style의 cell을 사용할 때,title과 subtitle의 변수명을 그대로 사용하였습니다. 이러한 방식이 괜찮은지 궁금합니다.

- title, subtitle 을 그대로 사용해도 괜찮지만 이름, 나이, 연락처를 나타낼 수 있도록 더 구체적인 이름으로 선언하는 방식이 더 좋을 것 같습니다!

이외에도 몇 가지 궁금증을 남겨보았습니다.
확인 부탁드립니다!

ohdair committed:
refactor: 프로퍼티 타입을 상수로 변경

zhunhe committed:
refactor: 접근제어자 private 설정

zhunhe committed:
refactor: ContactTableViewCell에 className 변수 구현

ohdair committed:
refactor: Error 채택된 부분을 삭제

zhunhe committed:
refactor: ContactTableViewCell에서 Cell 관련 데이터 설정

- configure(contact:) 구현
- IBOutlet 변수명 변경
title -> nameAndAge
subtitle -> phoneNumber

ohdair committed:
refactor: ContactTableViewCell 관련 수정

- 상속 불필요 클래스 final 키워드 추가 (메서드 디스패치 최적화)
- 내부 속성의 접근 제어자 private 추가

zhunhe committed:
fix: 강제 언래핑 사용하지 않도록 수정

zdodev commented on ContactManager/ContactManagerUI/Controller/ViewController.swift:
```diff
@@ -0,0 +1,61 @@
+        ) as? ContactTableViewCell else {
+            return UITableViewCell()
+        }
+        cell.configure(contact: contacts[indexPath.row])
```
좋습니다!👍

zdodev reviewed (approved):
안녕하세요. @ohdair , @zhunhe !

질문에 대해 좋은 답변 잘 달아주셨습니다!👍
코드도 적절하게 잘 개선해주셨네요!

Step 1 & 2는 여기서 마무리하도록 하겠습니다!🎉
